File: source/zc95/python/lib/ZcWs.py
import websocket # pip3 install websocket-client
import threading
import json 
import sys
import queue

class ZcWs:
  def __init__(self, ip, rcv_queue, debug):
    connection_string = "ws://" + ip + "/stream"
    
    print("Connecting")   
    self.ws = websocket.WebSocketApp(connection_string,
                              on_open    = self.__on_open,
                              on_message = self.__on_message,
                              on_error   = self.__on_error,
                              on_close   = self.__on_close)
    
    self.__recv_waiting = False
    self.__pending_recv_message = ""
    self.__recv_event = threading.Event()
    self.__waiting_for_msgId = 0
    self.__connection_wait_event = threading.Event()
    self.__rcv_queue = rcv_queue
    self.debug = debug
  
  def __on_message(self, ws, message):
    if self.debug:
      print("< " + message)
      
    result = json.loads(message)
    if self.__recv_waiting and "MsgId" in result and result["MsgId"] == self.__waiting_for_msgId:
        self.__pending_recv_message = message
        self.__recv_event.set()
    else:
      self.__rcv_queue.put(result)
    
  def __on_error(self, ws, error):
    if str(error) != "None" and len(str(error)) > 1:
      print("Websocket error: " + str(error))

  def __on_close(self, ws, close_status_code, close_msg):
    print("Websocket connection closed")
    sys.exit()

  def __on_open(self, ws):
    print("Connection opened")
    self.__connection_wait_event.set()
    
  def wait_for_connection(self):
    self.__connection_wait_event.wait()
    
  def send(self, message):
    if self.debug:
      print("> " + message)
    self.ws.send(message)
    
  # note: not at all thread safe
  def recv(self, msgId):
    self.__waiting_for_msgId = msgId
    self.__recv_waiting = True
    
    if self.__recv_event.wait(timeout=2): # timeout is seconds
      retval = self.__pending_recv_message
    else:
      retval = None
    
    self.__recv_waiting = False
    self.__recv_event.clear()
    self.__waiting_for_msgId = 0
    
    return retval
  
  def run_forever(self):
    self.ws.run_forever(ping_interval=6)
    
  def stop(self):
    self.ws.close()

File: misc/Arduino/libraries/Zc95/src/ZcWs.cpp
#include "ZcWs.h"

ZcWs::ZcWs(std::string ip, std::queue<std::string> &rcvQueue, bool debug) : _rcvQueue(rcvQueue)
{
  _debug = debug;
  _waitingForMsgId = 0;
  _recvWaiting = false;
  _ip = ip;
}

bool ZcWs::connect()
{
  _ws.onMessage(std::bind(&ZcWs::on_message, this, std::placeholders::_1));
  _ws.onEvent(std::bind(&ZcWs::on_event, this, std::placeholders::_1, std::placeholders::_2));
  if (_ws.connect(_ip.c_str(), 80, "/stream"))
  {
    Serial.println("Connected to WS");
    return true;
  }
  else
  {
    Serial.println("Failed to connect to WS");
    return false;
  }
}

void ZcWs::wait_for_connection()
{
  /*
  while (!_ws.available())
  {
    _ws.poll();
    delay(10);
  }
  */
}

void ZcWs::send(std::string message)
{
  if (_debug)
  {
    Serial.print("> ");
    Serial.println(message.c_str());
  }
  _ws.send(message.c_str());
  _next_ping_time = millis() + 4000;
}

// Wait up to 2s for an inbound message with the supplied messageId
std::string ZcWs::recv(int msgId)
{
  _waitingForMsgId = msgId;

  if (_pendingRecvMessage == "")
    _recvWaiting = true;

  unsigned long startTime = millis();
  while (millis() - startTime < 2000)
  {
    _ws.poll();
    if (_pendingRecvMessage != "")
    {
      std::string retval = _pendingRecvMessage;
      _pendingRecvMessage = "";
      _recvWaiting = false;
      return retval;
    }
    delay(10);
  }
  _recvWaiting = false;
  return "";
}

void ZcWs::loop()
{
  // The zc95 will close the connection after a few seconds of inactivity, so send
  // websocket pings if nothing else is being sent.
  if (_connected && millis() > _next_ping_time)
  {
    _ws.ping("");
    _next_ping_time = millis() + 4000;
  }

  _ws.poll();
}

void ZcWs::disconnect() { _ws.close(); }

bool ZcWs::is_connected() { return _connected; }

void ZcWs::on_message(websockets::WebsocketsMessage message)
{
  _next_ping_time = millis() + 4000;
  std::string msg = message.data().c_str();
  if (_debug)
  {
    Serial.print("< ");
    Serial.println(msg.c_str());
  }

  DynamicJsonDocument doc(1024);
  deserializeJson(doc, msg);
  JsonObject result = doc.as<JsonObject>();

  if (_recvWaiting && result.containsKey("MsgId") && result["MsgId"] == _waitingForMsgId)
  {
    _pendingRecvMessage = msg;
    _recvWaiting = false;
  }
  else
  {
    if (_rcvQueue.size() < MAX_RECV_QUEUE_LENGTH)
      _rcvQueue.push(msg);
    else if (_debug)
      Serial.println("Queue full, disgarding message"); // Not all programs will care about update messages, so this likely isn't an error
  }
}

void ZcWs::on_event(websockets::WebsocketsEvent event, String data)
{
  if (event == websockets::WebsocketsEvent::ConnectionOpened)
  {
    Serial.println("Connnection Opened");
    _next_ping_time = millis() + 4000;
    _connected = true;
  }
  else if (event == websockets::WebsocketsEvent::ConnectionClosed)
  {
    Serial.println("Connnection Closed");
    _connected = false;
  }
  else if (event == websockets::WebsocketsEvent::GotPing)
  {
    Serial.println("Got a Ping!");
  }
  else if (event == websockets::WebsocketsEvent::GotPong)
  {
    Serial.println("Got a Pong!");
  }
}

void ZcWs::on_error(websockets::WebsocketsMessage error)
{
  /*
  std::string errMsg = error.data();
  if (errMsg != "None" && errMsg.length() > 1)
  {
    Serial.println("Websocket error: " + errMsg);
  }
  */
}
